File: netlify/functions/ai-breakdown.mjs
const jsonResponse = (status, body) => ({
  statusCode: status,
  headers: {
    "Content-Type": "application/json; charset=utf-8",
    "Cache-Control": "no-store"
  },
  body: JSON.stringify(body)
});

const parseModelJson = (content) => {
  const match = String(content || "").match(/\{[\s\S]*\}/);
  if (!match) throw new Error("Model JSON was not found");
  return JSON.parse(match[0]);
};

export const handler = async (event) => {
  if (event.httpMethod !== "POST") {
    return jsonResponse(405, { detail: "Method not allowed" });
  }

  let payload;
  try {
    payload = JSON.parse(event.body || "{}");
  } catch {
    return jsonResponse(400, { detail: "请求格式无效" });
  }

  const goal = String(payload.goal || "").trim();
  if (!goal || goal.length > 200) {
    return jsonResponse(422, { detail: "目标长度需要在 1 到 200 个字符之间" });
  }

  const apiKey = process.env.STARTIFY_LLM_API_KEY;
  const baseUrl = String(process.env.STARTIFY_LLM_BASE_URL || "https://api.deepseek.com").replace(/\/$/, "");
  const model = process.env.STARTIFY_LLM_MODEL || "deepseek-v4-flash";
  const timeoutMs = Math.max(1000, Number(process.env.STARTIFY_LLM_TIMEOUT_SECONDS || 30) * 1000);
  if (!apiKey || !baseUrl || !model) {
    return jsonResponse(503, { detail: "在线大模型尚未配置" });
  }

  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const requestBody = {
      model,
      temperature: 0.2,
      max_tokens: 400,
      messages: [
        {
          role: "system",
          content: "你是 Startify 的任务启动助手。把模糊目标缩成一个 5 到 10 分钟内能开始的动作。只返回 JSON。"
        },
        {
          role: "user",
          content: `JSON 字段固定为 suggestionTitle、durationMin、energyLevel、reason、steps。\n目标：${goal}`
        }
      ]
    };
    if (baseUrl.includes("api.deepseek.com")) {
      requestBody.thinking = { type: "disabled" };
      requestBody.response_format = { type: "json_object" };
    }

    const response = await fetch(`${baseUrl}/chat/completions`, {
      method: "POST",
      signal: controller.signal,
      headers: {
        "Authorization": `Bearer ${apiKey}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify(requestBody)
    });
    if (!response.ok) {
      return jsonResponse(502, { detail: "大模型服务暂时不可用" });
    }

    const providerPayload = await response.json();
    const parsed = parseModelJson(providerPayload?.choices?.[0]?.message?.content);
    const steps = Array.isArray(parsed.steps)
      ? parsed.steps.map((step) => String(step).trim()).filter(Boolean).slice(0, 5)
      : [];
    if (!parsed.suggestionTitle || !parsed.reason || steps.length === 0) {
      throw new Error("Model payload is incomplete");
    }

    return jsonResponse(200, {
      goal,
      suggestionTitle: String(parsed.suggestionTitle).slice(0, 200),
      durationMin: Math.min(30, Math.max(1, Number(parsed.durationMin) || 5)),
      energyLevel: Math.min(3, Math.max(1, Number(parsed.energyLevel) || 1)),
      reason: String(parsed.reason),
      steps,
      source: `api:${model}`
    });
  } catch (error) {
    console.error("Startify AI breakdown failed", error);
    return jsonResponse(502, { detail: "大模型返回格式无效" });
  } finally {
    clearTimeout(timeoutId);
  }
};

File: backend/app/services/ai_breakdown.py
from __future__ import annotations

import json
import re
from typing import Any

import httpx

from ..config import Settings
from ..schemas import AIBreakdownResponse


class AIProviderError(RuntimeError):
    pass


def build_mock_breakdown(goal: str) -> AIBreakdownResponse:
    compact_goal = goal.strip()
    return AIBreakdownResponse(
        goal=compact_goal,
        suggestion_title=f"先处理：{compact_goal[:18]}",
        duration_min=5,
        energy_level=1,
        reason="当前后端没有配置大模型 API，使用透明标注的规则演示结果。",
        steps=[
            f"把“{compact_goal}”缩成一个 5 分钟内能开始的动作。",
            "先打开相关资料或工作区。",
            "只完成第一步，再决定是否继续。",
        ],
        source="mock",
    )


async def generate_breakdown(goal: str, settings: Settings) -> AIBreakdownResponse:
    compact_goal = goal.strip()
    if not settings.llm_configured:
        return build_mock_breakdown(compact_goal)

    endpoint = f"{settings.llm_base_url}/chat/completions"
    request_payload = {
        "model": settings.llm_model,
        "temperature": 0.2,
        "max_tokens": 400,
        "messages": [
            {
                "role": "system",
                "content": (
                    "你是 Startify 的任务启动助手。你的目标是降低启动门槛，"
                    "把模糊目标缩成一个 5 到 10 分钟内能开始的具体动作。"
                    "只返回 JSON，不要输出 Markdown。"
                ),
            },
            {
                "role": "user",
                "content": (
                    "请拆解下面的目标。JSON 字段固定为："
                    '{"suggestionTitle": string, "durationMin": integer, '
                    '"energyLevel": integer, "reason": string, "steps": string[]}。'
                    f"\n目标：{compact_goal}"
                ),
            },
        ],
    }
    if "api.deepseek.com" in settings.llm_base_url:
        request_payload["thinking"] = {"type": "disabled"}
        request_payload["response_format"] = {"type": "json_object"}

    try:
        async with httpx.AsyncClient(timeout=settings.llm_timeout_seconds) as client:
            response = await client.post(
                endpoint,
                headers={
                    "Authorization": f"Bearer {settings.llm_api_key}",
                    "Content-Type": "application/json",
                },
                json=request_payload,
            )
            response.raise_for_status()
            provider_payload = response.json()
    except (httpx.HTTPError, ValueError) as exc:
        raise AIProviderError("大模型服务暂时不可用，请稍后重试。") from exc

    try:
        content = provider_payload["choices"][0]["message"]["content"]
        parsed = _parse_json_object(content)
        steps = [str(step).strip() for step in parsed["steps"] if str(step).strip()][:5]
        if not steps:
            raise ValueError("steps is empty")
        return AIBreakdownResponse(
            goal=compact_goal,
            suggestion_title=str(parsed["suggestionTitle"]).strip(),
            duration_min=int(parsed["durationMin"]),
            energy_level=int(parsed["energyLevel"]),
            reason=str(parsed["reason"]).strip(),
            steps=steps,
            source=f"api:{settings.llm_model}",
        )
    except (KeyError, TypeError, ValueError) as exc:
        raise AIProviderError("大模型返回格式不符合任务拆解协议。") from exc


def _parse_json_object(content: Any) -> dict[str, Any]:
    if not isinstance(content, str):
        raise ValueError("content is not text")
    match = re.search(r"\{[\s\S]*\}", content)
    if not match:
        raise ValueError("JSON object not found")
    payload = json.loads(match.group(0))
    if not isinstance(payload, dict):
        raise ValueError("payload is not an object")
    return payload
